### PR TITLE
feat: attest Policy 2.1 execution capabilities

### PR DESCRIPTION
**Pull request by dacoldest**
## Summary
- add typed Policy 2.1 capability and execution-grant claims
- bind execution grants to the attested policy hash and enforce their short-lived validity window
- preserve the current hybrid PQC documentation from main

## Verification
- `npm run typecheck`
- `npm test` (12 tests)
- `npm run build`

## Summary by CodeRabbit

* **New Features**
  * Added support for `execution_grant` and execution capability claims (`capabilities`) in intent attestation verification.
* **Bug Fixes**
  * Strengthened intent attestation validation for `policyHash`, tightened `iat/exp` timing rules with clock tolerance, enforced execution-grant window constraints, and added cross-checking between attestation and its grant.
* **Documentation**
  * Updated the intent-attestation overview and normative verification rules for `execution_grant` and capability requirements.
* **Tests**
  * Expanded test coverage for valid/invalid grants, hash mismatches, and tolerance/lifetime boundary cases.
* **Chores**
  * Re-exported the `ExecutionGrantClaim` type and updated the verified claims shape to require `iat`.

### CODE REVIEW

File: src/verify.ts [hunk 1]
```diff
@@ -43,6 +45,18 @@ function isIntent(value: unknown): value is Intent {
...
+  if (typeof value !== "object" || value === null) return false;
+  const candidate = value as Record<string, unknown>;
+  const strings = ["policy_hash", "manifest_sha256", "shim_id", "executor_id", "adapter", "adapter_version", "nonce"];
+  if (strings.some((key) => typeof candidate[key] !== "string" || candidate[key] === "")) return false;
+  if (!Number.isSafeInteger(candidate.issued_at) || !Number.isSafeInteger(candidate.expires_at)) return false;
+  const issuedAt = candidate.issued_at as number;
+  const expiresAt = candidate.expires_at as number;
+  if (expiresAt <= issuedAt || expiresAt - issuedAt > 300) return false;
+  return candidate.repository_id === undefined || typeof candidate.repository_id === "string";
+}
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 2]
```diff
@@ -43,6 +45,18 @@ function isIntent(value: unknown): value is Intent {
   return true;
 }
 
+function isExecutionGrant(value: unknown): value is ExecutionGrantClaim {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 3]
```diff
@@ -43,6 +45,18 @@ function isIntent(value: unknown): value is Intent {
...
+  const strings = ["policy_hash", "manifest_sha256", "shim_id", "executor_id", "adapter", "adapter_version", "nonce"];
+  if (strings.some((key) => typeof candidate[key] !== "string" || candidate[key] === "")) return false;
+  if (!Number.isSafeInteger(candidate.issued_at) || !Number.isSafeInteger(candidate.expires_at)) return false;
+  const issuedAt = candidate.issued_at as number;
+  const expiresAt = candidate.expires_at as number;
+  if (expiresAt <= issuedAt || expiresAt - issuedAt > 300) return false;
+  return candidate.repository_id === undefined || typeof candidate.repository_id === "string";
+}
+
 export async function verifyIntentAttestation(
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 4]
```diff
@@ -107,6 +121,29 @@ export async function verifyIntentAttestation(
...
+  if (isExecutionGrant(payload.execution_grant)) {
+    const now = Math.floor(Date.now() / 1000);
+    if (
+      payload.execution_grant.expires_at <= now ||
+      payload.execution_grant.issued_at > now + CLOCK_TOLERANCE_SECONDS
+    ) {
+      throw new InvalidPayloadError("execution_grant is expired or not yet valid");
+    }
+  }
+  if (payload.capabilities !== undefined && (!Array.isArray(payload.capabilities) || payload.capabilities.some((capability) => typeof capability !== "string" || capability.length === 0))) {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 5]
```diff
@@ -43,6 +54,100 @@ function isIntent(value: unknown): value is Intent {
...
+  const now = Math.floor(Date.now() / 1000);
+  if (
+    grant.expires_at <= now ||
+    grant.issued_at > now + CLOCK_TOLERANCE_SECONDS
+  ) {
+    throw new InvalidPayloadError("execution_grant is expired or not yet valid");
+  }
+};
+
+const validatePolicyClaims = (
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 6]
```diff
@@ -43,6 +54,100 @@ function isIntent(value: unknown): value is Intent {
...
+  }
+  validateExecutionGrantWindow(executionGrant);
+  return {
+    policyHash: payload.policyHash,
+    capabilities,
+    executionGrant,
+  };
+};
+
 export async function verifyIntentAttestation(
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 7]
```diff
@@ -121,6 +228,14 @@ export async function verifyIntentAttestation(
       iss: payload.iss as string,
       exp: payload.exp as number,
       ...(iat !== undefined && { iat }),
+      policyHash: policyClaims.policyHash,
+      ...(typeof payload.scope === "string" && { scope: payload.scope }),
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 8]
```diff
@@ -32,73 +43,203 @@ function normalizeTrustedIssuers(
   }
 
   return normalized;
-}
+};
 
-function isIntent(value: unknown): value is Intent {
+const isIntent = (value: unknown): value is Intent => {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 9]
```diff
@@ -32,73 +43,203 @@ function normalizeTrustedIssuers(
...
+    value.some((capability) => !isNonEmptyString(capability))
+  ) {
+    throw new InvalidPayloadError(
+      "capabilities claim must be a list of non-empty strings"
+    );
+  }
+  return value;
+};
+
+const validateBoundExecutionGrant = (
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 10]
```diff
@@ -32,73 +43,203 @@ function normalizeTrustedIssuers(
...
       `Expected alg=EdDSA, got alg=${header.alg}`
     );
   }
+  return header;
+};
 
-  // Step 2: Verify signature + standard claims
-  let payload: Record<string, unknown>;
-  let protectedHeader: Record<string, unknown>;
+const throwVerificationError = (error: unknown): never => {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts:278]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 11]
```diff
@@ -32,73 +43,203 @@ function normalizeTrustedIssuers(
...
+    return {
+      payload: result.payload as Record<string, unknown>,
+      protectedHeader: result.protectedHeader as Record<string, unknown>,
+    };
+  } catch (error: unknown) {
+    return throwVerificationError(error);
   }
+};
+
+export const verifyIntentAttestation = async (
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts:349]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 12]
```diff
@@ -32,73 +49,271 @@ function normalizeTrustedIssuers(
...
-  // Step 2: Verify signature + standard claims
-  let payload: Record<string, unknown>;
-  let protectedHeader: Record<string, unknown>;
+const isIssuerClaimError = (
+  code: string | undefined,
+  message: string
+): boolean =>
+  code === "ERR_JWT_CLAIM_VALIDATION_FAILED" && /iss/i.test(message);
+
+const throwVerificationError = (error: unknown): never => {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: tests/verify.test.ts [hunk 13]
```diff
@@ -58,6 +109,340 @@ describe("verifyIntentAttestation", () => {
...
+  it("rejects a token issued beyond the five-second clock tolerance", async () => {
+    const { privateKey, jwks } = await makeEdDSAKeypairAndJWKS();
+    const jwt = await new SignJWT({
+      intent: VALID_INTENT,
+      policyHash: "policy-hash",
+    })
+      .setProtectedHeader({ alg: "EdDSA" })
+      .setIssuer("sigil-core")
+      .setExpirationTime("1h")
+      .setIssuedAt(Math.floor(Date.now() / 1000) + 6)
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply tests/verify.test.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 14]
```diff
@@ -32,73 +49,271 @@ function normalizeTrustedIssuers(
...
+
+const validateTemporalClaims = (payload: Record<string, unknown>): void => {
+  if (!Number.isSafeInteger(payload.exp)) {
+    throw new InvalidPayloadError("Payload missing or invalid exp");
+  }
+  if (!Number.isSafeInteger(payload.iat)) {
+    throw new InvalidPayloadError("Payload missing or invalid iat");
+  }
+  const now = Math.floor(Date.now() / 1000);
+  if ((payload.iat as number) > now + CLOCK_TOLERANCE_SECONDS) {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 15]
```diff
@@ -32,96 +49,300 @@ function normalizeTrustedIssuers(
...
+  if (code !== undefined && SIGNATURE_ERROR_CODES.has(code)) {
+    return new InvalidSignatureError();
+  }
+  return undefined;
+};
+
+const createVerificationError = (error: unknown): SigilVerificationError => {
+  if (error instanceof SigilVerificationError) return error;
+  const message = getErrorMessage(error);
+  const code = (error as { code?: string }).code;
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

File: src/verify.ts [hunk 16]
```diff
@@ -32,96 +50,309 @@ function normalizeTrustedIssuers(
...
+  const now = Math.floor(Date.now() / 1000);
+  if (
+    grant.expires_at <= now ||
+    grant.issued_at > now + CLOCK_TOLERANCE_SECONDS
+  ) {
+    throw new InvalidPayloadError("execution_grant is expired or not yet valid");
+  }
+};
+
+const validateTemporalClaims = (payload: Record<string, unknown>): void => {
```
[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.

[the remark this answers is not in the document]
Comment (dacoldest) [reply src/verify.ts]: Addressed in current head 1bb3479. The full local suite passes, and current-head CodeRabbit and DeepSource checks are green.
